### DISCUSSION TITLE
Config template - interfaces under tagged VLANs?

### DISCUSSION THREAD

**Discussion by etfz** [netbox-community/netbox#12569] [Help Wanted!]
Hi,

Looking to generate some VLAN configuration for MikroTik RouterOS. Unfortunately, the configuration required is a little backwards compared to the documentation, where you define a VLAN primarily, and then include relevants interfaces.

For example, if I have VLAN 5 tagged on sfp-sfpplus1, ether2 and ether3, and VLAN 10 tagged on ether2 and ether3, I need to generate the following configuration:
```
/interface bridge vlan add bridge=bridge vlan-ids=5 tagged=sfp-sfpplus1,ether2,ether3
/interface bridge vlan add bridge=bridge vlan-ids=10 tagged=ether2,ether3
```

Is there any sane way to accomplish this?

**Comment by candlerb** [comment]
The templating language [jinja2](https://jinja.palletsprojects.com/en/3.1.x/templates/) is quite powerful - it has [variables](https://jinja.palletsprojects.com/en/3.1.x/templates/#assignments), [looping](https://jinja.palletsprojects.com/en/3.1.x/templates/#for) etc.

You'd want to iterate over all interfaces, accumulate the list of VLANs seen; then iterate over each unique VLAN seen and find the interfaces which are a member of it (tagged or untagged). Here is something that comes close:

```
{%- for i in device.interfaces.all() %}
{%- for v in (i.tagged_vlans.all()|list) + [i.untagged_vlan] %}
{%- if v -%}{# and v not in seen #}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- set ifs = device.interfaces.filter(tagged_vlans=v)|list %}
{%- if ifs %} tagged={{ ifs|join(",") }}
{%- endif %}
{%- set ifs = device.interfaces.filter(untagged_vlan=v)|list -%}
{%- if ifs %} untagged={{ ifs|join(",") }}
{%- endif %}
{%- endif %}
{%- endfor %}
{%- endfor %}
```

Unfortunately, when it sees the same vlan multiple times, it repeats the config.  Fixing that seems to be tricky, because Netbox doesn't enable the ["do" extension](https://jinja.palletsprojects.com/en/3.1.x/templates/#expression-statement), which would allow you to update a list or a dict of vlans already seen. You could raise that as a feature request.

With Jinja2, it's also possible to add your own [custom filters](https://jinja.palletsprojects.com/en/3.1.x/api/#custom-filters) in Python, like [these](https://docs.netbox.dev/en/stable/plugins/development/templates/#filters) which Netbox itself provides - but I don't know if there's a supported way to add your own in Netbox, without writing a complete Plugin.

EDIT: I realised this doesn't handle interface setting "Tagged (all)". That's tricky; I think you'd need to get a list of all VLANs in the site. Actually, that suggests a solution for the first problem, as long as you've assigned each VLAN to the site:

```
{%- for v in device.site.vlans.all() %}
{%- set ift = device.interfaces.filter(tagged_vlans=v)|list %}
{%- set ifu = device.interfaces.filter(untagged_vlan=v)|list -%}
{%- if ift or ifu %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- if ift %} tagged={{ ift|join(",") }}{% endif %}
{%- if ifu %} untagged={{ ifu|join(",") }}{% endif %}
{%- endif %}
{%- endfor %}
```

It's a bit tricker to add in interfaces with tagged-all.  In Python/Django it's easy to combine querysets:

```
device.interfaces.filter(tagged_vlans=v) | device.interfaces.filter(mode="tagged-all")
```

But that conflicts with the pipe operator in Jinja2.  The following seems to work though:

```
{%- for v in device.site.vlans.all() %}
{%- set ift = ((device.interfaces.filter(tagged_vlans=v)|list) + (device.interfaces.filter(mode="tagged-all")|list)) | unique | sort(attribute="name") %}
{%- set ifu = device.interfaces.filter(untagged_vlan=v) | sort(attribute="name") -%}
{%- if ift or ifu %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- if ift %} tagged={{ ift|join(",") }}{% endif %}
{%- if ifu %} untagged={{ ifu|join(",") }}{% endif %}
{%- endif %}
{%- endfor %}
```

**Reply by kkthxbye-code** [reply]
> Fixing that seems to be tricky, because Netbox doesn't enable the ["do" extension](https://jinja.palletsprojects.com/en/3.1.x/templates/#expression-statement), which would allow you to update a list or a dict of vlans already seen.

https://github.com/netbox-community/netbox/issues/12378

> You can add environment parameters to the Config Template in the Environment params field.
`{"extensions": ["jinja2.ext.do"]}`

Regarding the overall issue, it seems similar to what I helped someone with on slack recently, maybe it can be of help:

https://netdev-community.slack.com/archives/C01P0FRSXRV/p1683533373634269

**Reply by etfz** [reply]
> EDIT: I realised this doesn't handle interface setting "Tagged (all)". That's tricky; I think you'd need to get a list of all VLANs in the site. Actually, that suggests a solution for the first problem, as long as you've assigned each VLAN to the site:

We mostly use VLAN groups rather than site assignments, but also I don't think we use tagged all anywhere so far, so we could do without for now. Untagged VLANs are also handled elsewhere. With that in mind, this is what I've got:

```
{% set vlans = [] %}
{% for interface in device.interfaces.all() -%}
    {% set _ = vlans.extend(interface.tagged_vlans.all()) -%}
{% endfor -%}

{%- for v in vlans | unique %}
{%- set ift = ((device.interfaces.filter(tagged_vlans=v)|list) + (device.interfaces.filter(mode="tagged-all")|list)) %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}} tagged={{ ift|join(',') }}
{%- endfor %}
```

> Regarding the overall issue, it seems similar to what I helped someone with on slack recently, maybe it can be of help:
>
> https://netdev-community.slack.com/archives/C01P0FRSXRV/p1683533373634269

Not too familiar with Slack, but I don't seem to have access to that community?

**Reply by candlerb** [reply]
Here's an alternative formulation thanks to @v0tti in #12573

```
{%- for v in device.site.vlans.all() %}
{%- set ift = ((v.interfaces_as_tagged.exclude(type="virtual").filter(device=device)|list)
             + (device.interfaces.filter(mode="tagged-all")|list))|sort(attribute="name") %}
{%- set ifu = v.interfaces_as_untagged.exclude(type="virtual").filter(device=device)|sort(attribute="name") %}
{%- if ift or ifu %}
/interface bridge vlan add bridge=bridge vlan-ids={{v.vid}}
{%- if ift %} tagged={{ ift|join(",") }}{% endif %}
{%- if ifu %} untagged={{ ifu|join(",") }}{% endif %}
{%- endif %}
{%- endfor %}
```

**Reply by dreng** [reply]
I'm experimenting with VLANs and config templates too. I started with this simple example:

```
{% for interface in device.interfaces.all() %}
  Interface ({{ interface.id }}): {{ interface }}
  Untagged VLANs: {{ interface.untagged_vlan }}
  Tagged VLANs: {% for vlan in interface.tagged_vlans.all() %}{{ vlan.vid }} {% endfor %}
{% endfor %}
```

The problem is that I set one of the interfaces in the test environment to "802.1Q Mode: Tagged (All)". I can't seem to find a way to query the tagged VLANs then. I would expect that list of all VLANs is returned. Instead, interface.tagged_vlans.all()  returns an empty list. I even can't query the 802.1Q Mode in order to work around this.

Am I missing something?

**Reply by candlerb** [reply]
> Am I missing something?

"Tagged (All)" is effectively just a status meaning "no list of VLANs provided".  In many switches, not configuring a list of VLANs on the port makes it participate in any or all VLANs that are there.

Netbox doesn't return all VLAN objects, because it doesn't know what set of VLANs is available to this particular device or rack or location.

If you want all VLANs in the site, you could use `device.site.vlans.all()` - assuming each VLAN has been bound directly to a site (i.e. it probably doesn't work if bound to a vlan group, but I haven't tested that). See [above](https://github.com/netbox-community/netbox/discussions/12569#discussioncomment-5885993) and #12573.

> I even can't query the 802.1Q Mode in order to work around this

The 802.1Q Mode is an enumerated value: "access", "tagged" or "tagged-all". In [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/):

```
>>> Device.objects.get(name="gw2").interfaces.get(name="ether5").mode
'tagged'
```

**Reply by dreng** [reply]
Thanks for clarifying! So, this will do the trick:

```
{% for interface in device.interfaces.all() %}
  Interface ({{ interface.id }}): {{ interface }}
  Untagged VLANs: {{ interface.untagged_vlan.vid }}
  Tagged VLANs: {% if interface.mode == 'tagged-all' %}{% for vlans in device.site.vlans.all() %}{{ vlans.vid }} {% endfor %}{% else %}{% for vlan in interface.tagged_vlans.all() %}{{ vlan.vid }} {% endfor %}{% endif %}
{% endfor %}
```

This is unfortunately horribly formatted, but otherwise the output looks awful.

**Reply by peterbaumert** [reply]
Maybe a "jinja templates" section in the device types github would be a good addon where people could share their examples. Or a completly own repo?
